streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -17,242 +17,116 @@
-        this.zkClient = new ZkClient(zkConnect, 30 * 1000, 30 * 1000, new ZKStringSerializer());
+    public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient, final int replicationFactor, final long windowChangeLogAdditionalRetention) {
+        this.streamsKafkaClient = streamsKafkaClient;
         this.replicationFactor = replicationFactor;
         this.windowChangeLogAdditionalRetention = windowChangeLogAdditionalRetention;
     }
 
-    public void makeReady(InternalTopicConfig topic, int numPartitions) {
-        boolean topicNotReady = true;
-
-        while (topicNotReady) {
-            Map<Integer, List<Integer>> topicMetadata = getTopicMetadata(topic.name());
+    /**
+     * Prepares the set of given internal topics. If the topic with the correct number of partitions exists ignores it. For the ones with different number of
+     * partitions delete them and create new ones with correct number of partitons along with the non existing topics.
+     * @param topic
+     */
+    public void makeReady(final InternalTopicConfig topic, int numPartitions) {
 
-            if (topicMetadata == null) {
-                try {
-                    createTopic(topic, numPartitions, replicationFactor);
-                } catch (ZkNodeExistsException e) {
-                    // ignore and continue
-                }
-            } else {
-                if (topicMetadata.size() > numPartitions) {
-                    // else if topic exists with more #.partitions than needed, delete in order to re-create it
-                    try {
-                        deleteTopic(topic.name());
-                    } catch (ZkNodeExistsException e) {
-                        // ignore and continue
-                    }
-                } else if (topicMetadata.size() < numPartitions) {
-                    // else if topic exists with less #.partitions than needed, add partitions
-                    try {
-                        addPartitions(topic.name(), numPartitions - topicMetadata.size(), replicationFactor, topicMetadata);
-                    } catch (ZkNoNodeException e) {
-                        // ignore and continue
-                    }
-                } else {
-                    topicNotReady = false;
-                }
+        Map<InternalTopicConfig, Integer> topics = new HashMap<>();
+        topics.put(topic, numPartitions);
+        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
+            try {
+                Collection<MetadataResponse.TopicMetadata> topicMetadatas = streamsKafkaClient.fetchTopicMetadata();
+                Map<InternalTopicConfig, Integer> topicsToBeDeleted = getTopicsToBeDeleted(topics, topicMetadatas);
+                Map<InternalTopicConfig, Integer> topicsToBeCreated = filterExistingTopics(topics, topicMetadatas);
+                topicsToBeCreated.putAll(topicsToBeDeleted);
+                streamsKafkaClient.deleteTopics(topicsToBeDeleted);
```

Comment:
This breaks backwards compatibility. Until now Streams did not delete existing topics. Until Streams offers a way to configure partition count and min.isr for internal topics, it should never attempt to delete topics. Even then it might be dangerous to delete existing topics without warning.